commands/upgrade_cluster.go

```diff
@@ -0,0 +1,311 @@
+	for _, r := range releasesResult {
+		releaseVersions = append(releaseVersions, *r.Version)
+	}
+
+	// define the target version to upgrade to
+	targetVersion := successorReleaseVersion(details.ReleaseVersion, releaseVersions)
+	if targetVersion == "" {
+		return result, microerror.Mask(noUpgradeAvailableError)
+	}
+
+	var targetRelease models.V4ReleaseListItem
+	for _, rel := range releasesResult {
+		if *rel.Version == targetVersion {
+			targetRelease = *rel
+		}
+	}
+
+	// Show some details independent of confirmation
+	if !targetRelease.Active {
+		fmt.Printf("Cluster '%s' will be upgraded from version %s to %s, which is not an active release.\n",
+			color.CyanString(args.clusterID),
+			color.CyanString(details.ReleaseVersion),
+			color.CyanString(targetVersion))
+		fmt.Printf("This might fail depending on your permissions.\n")
+	} else {
+		fmt.Printf("Cluster '%s' will be upgraded from version %s to %s.\n",
+			color.CyanString(args.clusterID),
+			color.CyanString(details.ReleaseVersion),
+			color.CyanString(targetVersion))
+	}
+
+	// Details output
+	fmt.Println("")
+	fmt.Println("Changelog:")
+	fmt.Println("")
+
+	for _, change := range targetRelease.Changelog {
+		fmt.Printf("    - %s: %s\n", change.Component, change.Description)
+	}
+
+	fmt.Println("")
+	fmt.Println("NOTE: Upgrading may impact your running workloads and will make the cluster's")
+	fmt.Println("Kubernetes API unavailable temporarily. Before upgrading, please acknowledge the")
+	fmt.Println("details described in")
+	fmt.Println("")
+	fmt.Printf("    %s\n", upgradeDocsURL)
+	fmt.Println("")
+
+	// Confirmation
+	if !args.force {
+		confirmed := askForConfirmation("Do you want to start the upgrade now?")
+		if !confirmed {
+			return result, microerror.Mask(commandAbortedError)
+		}
+	}
+
+	result.clusterID = args.clusterID
+	result.versionBefore = details.ReleaseVersion
+
+	// request body
+	reqBody := &models.V4ModifyClusterRequest{
+		ReleaseVersion: targetVersion,
+	}
+
+	// perform API call
+	auxParams := ClientV2.DefaultAuxiliaryParams()
+	auxParams.ActivityName = upgradeClusterActivityName
+	_, err := ClientV2.ModifyCluster(args.clusterID, reqBody, auxParams)
+	if err != nil {
+		return result, microerror.Maskf(couldNotUpgradeClusterError, err.Error())
+	}
+
+	result.versionAfter = targetVersion
+
+	return result, nil
+}
+
+// successorReleaseVersion returns the lowest version number from a slice
+// that is still higher than the comparison version.
+// If no successor is found, returns an empty string.
+func successorReleaseVersion(version string, versions []string) string {
+	// sort versions by semver number
+	sort.Slice(versions, func(i, j int) bool {
+		vi := semver.New(versions[i])
+		vj := semver.New(versions[j])
+		return vi.LessThan(*vj)
+	})
+
+	// return first item that is greater than version
+	comp := semver.New(version)
+	for _, v := range versions {
+		vv := semver.New(v)
+		if comp.LessThan(*vv) {
+			return v
+		}
+	}
+
+	return ""
```

Comment:
IMO here should an error be returned. Implying behaviour by a certain content difference is an anti pattern IMO.

Reply:
For me is a blocker also. Two tings: what if this is newest version? What if old cluster doesn't have a version (credits to Oli)? There are maybe even more corner cases. I prefer having error than random stuff happening.